cmd/swarm/swarm-smoke/pss.go

```diff
@@ -0,0 +1,213 @@
+	"github.com/ethereum/go-ethereum/swarm/pss"
+	"github.com/ethereum/go-ethereum/swarm/testutil"
+
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+const (
+	pssModeRaw = iota
+	pssModeAsym
+	pssModeSym
+)
+
+type pssJob struct {
+	msg  []byte
+	mode int
+	prox bool
+}
+
+type pssNode struct {
+	hostIdx        int
+	addr           []byte
+	pubkey         string
+	client         *rpc.Client
+	deregisterFunc func()
+	msgC           chan pss.APIMsg
+}
+
+// keyed by hex representation of overlay address
+type pssSession struct {
+	topic        pss.Topic
+	msgC         chan pss.APIMsg
+	nodes        []*pssNode
+	requiredJobs map[string]*pssJob
+	allowedJobs  map[string]*pssJob
+}
+
+func pssChecks(ctx *cli.Context, tuid string) error {
+	return pssAsym(ctx, tuid)
+}
+
+func pssAsym(ctx *cli.Context, tuid string) error {
+
+	errc := make(chan error)
+
+	session := pssSetup()
+	defer func() {
+		for _, n := range session.nodes {
+			n.deregisterFunc()
+		}
+	}()
+
+	go func() {
+		errc <- pssAsymDo(ctx, session, tuid)
+	}()
+
+	select {
+	case err := <-errc:
+		if err != nil {
+			metrics.GetOrRegisterCounter(fmt.Sprintf("%s.fail", commandName), nil).Inc(1)
+		}
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		metrics.GetOrRegisterCounter(fmt.Sprintf("%s.timeout", commandName), nil).Inc(1)
+
+		e := fmt.Errorf("timeout after %v sec", timeout)
+		return e
+	}
+
+}
+
+func pssSetup() *pssSession {
+
+	// random topic, one per session,  same for all msgs
+	topic := pss.BytesToTopic(testutil.RandomBytes(seed, 4))
+
+	session := &pssSession{
+		msgC:         make(chan pss.APIMsg),
+		requiredJobs: make(map[string]*pssJob),
+		allowedJobs:  make(map[string]*pssJob),
+		topic:        topic,
+	}
+
+	// hosts is global :/
+	// set up the necessary info for each pss node
+	for i, host := range hosts {
+		httpHost := fmt.Sprintf("ws://%s:%d", host, 8546)
+		rpcClient, err := rpc.Dial(httpHost)
+		if err != nil {
+			log.Error("Error dialing host", "err", err)
+			continue
+		}
+
+		// get overlay address
+		// TODO this should be done automatically for all nodes anyway in all smokes perhaps?
+		var addr hexutil.Bytes
+		err = rpcClient.Call(&addr, "pss_baseAddr")
+		if err != nil {
+			log.Error("Error calling host for addr", "err", err)
+			continue
+		}
+
+		// get pss public key
+		var pubkey string
+		err = rpcClient.Call(&pubkey, "pss_getPublicKey")
+		if err != nil {
+			log.Error("Error calling host for pubkey", "err", err)
+			continue
+		}
+
+		// subscribe to the topic for the session
+		// this creates the incoming message handler automatically
+		// any message received with this topic comes on the channel
+		//msgC := make(chan pss.APIMsg)
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		sub, err := rpcClient.Subscribe(ctx, "pss", session.msgC, "receive", session.topic, true, false)
+		if err != nil {
+			log.Error("Error calling host for subscribe", "err", err)
+			continue
```

Comment:
should we continue or fail the whole test?
what about metrics? I'd like to see a counter being incremented here

Reply:
I've update the test to send multiple messages (configurable via CLI flag `pss-messages`) via PSS using asymetric encryption between random nodes.

I've added the following counters:

`pss.asym.failMsg` - total failed messages
`pss.asym.successMsg` - total successful messages